### PR TITLE
[fix] fix and refactor TestValidateStatefulSet and TestValidateStatefulSet test cases

### PR DESCRIPTION
**Pull request by areller**
#### What type of PR is this?

bug
test false positive

#### What this PR does / why we need it:

Fixing a false positive in the `/apis/apps` validation test suite

#### Special notes for your reviewer:

I tried adding a new error case to [TestValidateStatefulSetUpdate](https://github.com/kubernetes/kubernetes/blob/9ac6b3af5f2015530627d3362e35712393fb6887/pkg/apis/apps/validation/validation_test.go#L751) for a different feature I'm working on.  
I then discovered that all the test cases in `errorCases` succeed by default because we don't set the `ResourceVersion` like we do for the `successCases` https://github.com/kubernetes/kubernetes/blob/9ac6b3af5f2015530627d3362e35712393fb6887/pkg/apis/apps/validation/validation_test.go#L916-L917

After setting `ResourceVersion` for `errorCases`, I found out that the `empty pod creation policy` fails because there is no validation on the pod creation policy being empty (or nil). I'm not sure if that's something that's wanted. If so, I can return the test and implement the appropriate validation. Please let me know

Does this PR introduce a user-facing change?
```
NONE
```

Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```
```

## Edit

* As part of @soltysh's suggestions, I'm now validating that the expected errors are returned (not only that there are errors) (see a31cbb5fb4d05ca8f188c557134d1fbe63816a68 and 62f7f55bfc9d52be98fb4f9f505d2edfa3f6f2f1) and I've also cleaned up some of the tests in `TestValidateStatefulSet` to make sure that one test case doesn't test too many things at once (e.g. a test that's called `invalid labels` shouldn't also test that `restartPolicy` is validated) (see b2a2ec8130ca0de62c75d843143b855a3bae9130) and cleaned up tests in `TestValidateStatefulSetUpdate` to make sure that it only tests validation of allowed/disallowed transitions (e.g. testing that changing the number of replica is allowed and changing PVC template isn't. not testing stuff like invalid labels that are already covered by `TestValidateStatefulSet`).
* As part of @liggitt suggestions, I've changed the commit history to make the changes more clear and I've modified `ValidateStatefulSetUpdate` to also validate that the new Statefulset is valid so that `stategy.go#ValidateUpdate` only has to call `ValidateStatefulSetUpdate`

### CODE REVIEW

File: pkg/apis/apps/validation/validation_test.go [hunk 1]
```diff
@@ -1044,6 +1025,8 @@ func TestValidateStatefulSetUpdate(t *testing.T) {
 
 	for testName, errorCase := range errorCases {
 		t.Run(testName, func(t *testing.T) {
+			errorCase.old.ObjectMeta.ResourceVersion = "1"
```
Comment (soltysh) [inline pkg/apis/apps/validation/validation_test.go:1028]: Good catch, can you change the current test to actually verify the contents of the errors instead of just their presences, something similar to https://github.com/kubernetes/kubernetes/blob/30adcd0b6c1c28303d7281b7be2ac45bdd7989e4/pkg/apis/batch/validation/validation_test.go#L737-L743 ?

**Review by soltysh — CHANGES_REQUESTED** [review]
/ok-to-test
/kind bug
/priority important-longterm
/triage accepted

File: pkg/apis/apps/validation/validation_test.go (hunk 1 above)
Comment (areller) [reply pkg/apis/apps/validation/validation_test.go]: @soltysh upon farther inspection, found out that most of the negative tests in `TestValidateStatefulSetUpdate` are misleading. for example, https://github.com/kubernetes/kubernetes/blob/8fb2b901adce763853103dbebd56acfc827a1e84/pkg/apis/apps/validation/validation_test.go#L984-L1003 doesn't fail because of invalid selectors/labels, it fails because you're only allowed to modify `template`, `replicas`, etc.. in a statefulset. same with any other negative test in `TestValidateStatefulSetUpdate`. in fact, it looks like `ValidateStatefulSetUpdate` doesn't care about invalid labels or most of the other stuff that happen during the validation of a statefulset, it mostly verifies that during the modification of a statefulset, you didn't modify stuff that you aren't allowed to modify (e.g. PVC template)
(mostly, because it does verify that replicas and some other fields are not negative, not sure what's the reason behind that).

So I've modified the tests in `TestValidateStatefulSetUpdate` to be more explicit in what they're testing (e.g. `update pod management policy should fail`)

I've applied your suggestion (verifying the contents of the errors) to both `TestValidateStatefulSetUpdate` and `TestValidateStatefulSet`

Comment (areller) [reply pkg/apis/apps/validation/validation_test.go]: I've also changed some of the tests in `TestValidateStatefulSet` to fail for one (or few) reasons instead of multiple unrelated reasons

for example, `negative_replicas` was failing both because `replicas` was negative but also because it was missing the pod template

Comment (soltysh) [reply pkg/apis/apps/validation/validation_test.go]: This look so much better, thanks a lot for making the test much better!

**Review by soltysh — APPROVED** [review]
/lgtm
/label tide/merge-method-squash
/assign @liggitt 
for approval

File: pkg/apis/apps/validation/validation_test.go (hunk 1 above)
Comment (liggitt) [reply pkg/apis/apps/validation/validation_test.go]: Setting resource version to pass update validation is a good catch.

> it looks like `ValidateStatefulSetUpdate` doesn't care about invalid labels or most of the other stuff that happen during the validation of a statefulset

hmm... that's pretty misleading, in my opinion. The only place we call ValidateStatefulSetUpdate (outside of tests) is from statefulSetStrategy#ValidateUpdate, where we also call ValidateStatefulSet.

It would be clearer and less error-prone if we moved the call to ValidateStatefulSet inside ValidateStatefulSetUpdate, rather than requiring two validation calls to get correct behavior, rather than removing/relaxing validation test cases like "empty pod creation policy" above.

I like the change to include the expected errors in the testcase, rather than just requiring failures to return _some_ error... that would have caught these scenarios failing with an error other than what we thought we were exercising.

Is there a way to separate out the whitespace-only non-functional test changes (like making the testcases a slice of named tests instead of a map) to their own commits so we can easily see the diffs that are changing test behavior?

File: pkg/apis/apps/validation/validation_test.go [hunk 2]
```diff
@@ -749,11 +909,12 @@ func TestValidateStatefulSetStatus(t *testing.T) {
 }
 
 func TestValidateStatefulSetUpdate(t *testing.T) {
-	validLabels := map[string]string{"a": "b"}
```
Comment (liggitt) [inline pkg/apis/apps/validation/validation_test.go:752]: why the change from validLabels to validLabels1?

File: pkg/apis/apps/validation/validation_test.go (hunk 1 above)
Comment (areller) [reply pkg/apis/apps/validation/validation_test.go]: @liggitt 

> I like the change to include the expected errors in the testcase, rather than just requiring failures to return some error... that would have caught these scenarios failing with an error other than what we thought we were exercising.
> 
> Is there a way to separate out the whitespace-only non-functional test changes (like making the testcases a slice of named tests instead of a map) to their own commits so we can easily see the diffs that are changing test behavior?

I can replace the last commit with 3

1. making the tests a slice instead of a map with the ability to provide expected errors (at this point, I won't provide any expected errors)
2. changing the actual tests (e.g. adding `update pvc template storage class`, removing `invalid pod`)
3. adding the expected errors to the error cases

would that make it better?

> hmm... that's pretty misleading, in my opinion. The only place we call ValidateStatefulSetUpdate (outside of tests) is from statefulSetStrategy#ValidateUpdate, where we also call ValidateStatefulSet.
> 
> It would be clearer and less error-prone if we moved the call to ValidateStatefulSet inside ValidateStatefulSetUpdate, rather than requiring two validation calls to get correct behavior, rather than removing/relaxing validation test cases like "empty pod creation policy" above.

I wouldn't mind, but does it mean that we'll need to duplicate some tests, e.g. testing invalid labels both in `ValidateStatefulSet` and `ValidateStatefulSetUpdate`?

Comment (liggitt) [reply pkg/apis/apps/validation/validation_test.go]: > 1. making the tests a slice instead of a map with the ability to provide expected errors (at this point, I won't provide any expected errors)

yes, no-op structural change

> 2. changing the actual tests (e.g. adding update pvc template storage class, removing invalid pod)
> 3. adding the expected errors to the error cases

adding the expected errors first would be clearer to understand what coverage we're actually getting today

> I wouldn't mind, but does it mean that we'll need to duplicate some tests, e.g. testing invalid labels both in ValidateStatefulSet and ValidateStatefulSetUpdate?

no, not really... keeping the update tests focused on disallowed transitions is sufficient

**Comment by liggitt** [comment]
This looks really good, just a few comments. You'll also need to rebase on master so this PR doesn't contain a merge commit.

File: pkg/apis/apps/validation/validation_test.go [hunk 3]
```diff
@@ -1405,7 +1406,7 @@ func TestValidateStatefulSetUpdate(t *testing.T) {
 			testCase.old.ObjectMeta.ResourceVersion = "1"
 			testCase.update.ObjectMeta.ResourceVersion = "1"
```
[the remark this answers is not in the document]
Comment (areller) [reply pkg/apis/apps/validation/validation_test.go:1408]: yeah. looks like this is what `strategy.go` uses. updated

File: pkg/apis/apps/validation/validation_test.go [hunk 4]
```diff
@@ -921,40 +923,85 @@ func TestValidateStatefulSetUpdate(t *testing.T) {
...
+				Requests: api.ResourceList{
+					api.ResourceStorage: resource.MustParse("1Gi"),
+				},
 			},
-			ObjectMeta: metav1.ObjectMeta{
-				Labels: invalidLabels,
+		},
+	}
+
+	validPVCTemplateChangedSize := api.PersistentVolumeClaim{
```
[the remark this answers is not in the document]
Comment (areller) [reply pkg/apis/apps/validation/validation_test.go:957]: love this trick :)

**Comment by liggitt** [comment]
thanks for the cleanup!
